spec/client/ics-028-wasm-client/README.md

```diff
@@ -0,0 +1,451 @@
+---
+ics: 28
+title: WASM Client
+stage: draft
+category: IBC/TAO
+kind: instantiation
+implements: 2
+author: Parth Desai <parth@chorus.one>
+created: 2020-10-13
+modified: 2020-10-13
+---
+
+## Synopsis
+
+This specification document describes an interface to a client (verification algorithm) stored as a wasm bytecode for a blockchain.
+
+### Motivation
+
+WASM based light clients are decoupled from SDK source code, which enables one to upgrade existing light client or add support for other blockchain without modifying SDK source code.
+
+### Definitions
+
+Functions & terms are as defined in [ICS 2](../ics-002-client-semantics).
+
+`currentTimestamp` is as defined in [ICS 24](../ics-024-host-requirements).
+
+`WASM Client Code` refers to wasm bytecode stored in the client store, which provides target blockchain specific implementation of [ICS 2](../ics-002-client-semantics).
+
+`WASM Client` refers to a particular instance of `WASM Client Code` defined as a tuple `(WASM Client Code, ClientID)`.
+
+`WASM VM` refers to a virtual machine capable of executing valid wasm bytecode.
+
+
+### Desired Properties
+
+This specification must satisfy the client interface defined in ICS 2.
+
+## Technical Specification
+
+This specification depends on correctness of the `WASM Client code` in context of consensus algorithm of its target `blockchain`, as well as correct instantiation of `WASM Client`.
+
+### Client state
+
+The wasm client state tracks location of the wasm bytecode via `codeId`. Binary data represented by `data` field is opaque and only interpreted by the WASM Client Code. `type` represents client type.
+`type` and `codeId` both are immutable.
+
+```typescript
+interface ClientState {
+  codeId: []byte
+  data: []byte
+  frozen: boolean
+  frozen_height: Height
+  latest_height: Height
+  type: String
+}
+```
+
+### Consensus state
+
+The WASM consensus state tracks the timestamp (block time), `WASM Client code` specific fields and commitment root for all previously verified consensus states.
+`type` and `codeId` both are immutable.
+
+```typescript
+interface ConsensusState {
```

Comment:
ditto, same question, why can't this all be up to the WASM code?

Reply:
@cwgoes See my previous comment.

Reply:
See my previous response

Reply:
These are methods without access to storage:

### Consensus State ([link](https://github.com/cosmos/ibc-go/blob/main/modules/core/exported/client.go#L188)):
```
ClientType() string
GetRoot() Root
GetTimestamp() uint64
ValidateBasic() error
```

Reply:
`GetRoot() Root`: This is only getting used in internal client logic, we could remove it altogether if desired.

`GetTimestamp() uint64`: We need the timestamp of the consensus state to process packet timeouts. We can give you the client store here. The function you posted is merely a convenience, we could use the local time instead.

`ValidateBasic() error`: Again, this must be a stateless check. You can do very basic checks (e.g. codeId not empty), or even a no-op, and perform your stateful checks later on in message execution when you have access to the store.

Reply:
> GetRoot() Root: This is only getting used in internal client logic, we could remove it altogether if desired.

Yes, if `GetRoot` is removed from the interface, then there is no need for us to require `root: MerkleRoot` field anymore.

> GetTimestamp() uint64: We need the timestamp of the consensus state to process packet timeouts. We can give you the client store here. The function you posted is merely a convenience, we could use the local time instead.

Great, then we can modify the function `modules/apps/transfer/client/cli/tx.go` and the only place where `GetTimeout()` is called is [03-connection/keeper/keeper.go](https://github.com/cosmos/ibc-go/blob/2523fdbfd8fc14554d81dc92ce58b3a590964647/modules/core/03-connection/keeper/keeper.go#L99) and here where we have access to the VM.

Similar question to `ZeroCustomFields`... To pull the timestamp from the Wasm light client we would have to call the VM, which may fail. The current interface `GetTimestamp() uint64` doesn't allow to return error, so would it be alright to change the signature to `GetTimestamp() (uint64, error)`?

Reply:
> GetRoot() Root: This is only getting used in internal client logic, we could remove it altogether if desired.

Yes, if `GetRoot` is removed from the interface, then there is no need for us to require `root: MerkleRoot` field anymore.

> GetTimestamp() uint64: We need the timestamp of the consensus state to process packet timeouts. We can give you the client store here. The function you posted is merely a convenience, we could use the local time instead.

Great, then we can modify the function `modules/apps/transfer/client/cli/tx.go` and the only place where `GetTimeout()` is called is [03-connection/keeper/keeper.go](https://github.com/cosmos/ibc-go/blob/2523fdbfd8fc14554d81dc92ce58b3a590964647/modules/core/03-connection/keeper/keeper.go#L99) and here where we have access to the VM. 

To pull the timestamp from the Wasm light client we would have to:
* call the VM, which may fail. The current interface `GetTimestamp() uint64` doesn't allow to return an error, so would it be alright to change the signature to `GetTimestamp() (uint64, error)`?
* `ConsensusState` doesn't hold `codeId` that is required to call the WasmVM. So far, it was only stored by `ClientState`, so here are the possible options:
  * change the signature `ConsensusState.GetTimestamp() uint64` -> `ConsensusState.GetTimestamp(cs ClientState) uint64` but that looks weird
  * add the `codeId` field to `ConsensusState`
  * other options?

Reply:
updated `ConsensusState` [here](https://github.com/ParthDesai/ibc/commit/d7c33fc5c01ff4c07000f89ff9da04c31c27ed23#diff-081ac6bcbed0e39a406e9c83aa79a058605a646a4047d5be67cb84dd9b83877bR62)